numpy/core/function_base.py

```diff
@@ -123,21 +123,43 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
     if num < 0:
         raise ValueError("Number of samples, %s, must be non-negative." % num)
     div = (num - 1) if endpoint else num
-
-    # Convert float/complex array scalars to float, gh-3504
-    # and make sure one can use variables that have an __array_interface__, gh-6634
-    start = asanyarray(start) * 1.0
-    stop  = asanyarray(stop)  * 1.0
+    # Make sure one can use variables that have an __array_interface__, gh-6634
+    start = asanyarray(start)
+    stop  = asanyarray(stop)
+
+    # check if dealing with datetime64 or timedelta64
+    is_datelike = start.dtype.kind in 'Mm'
+    if is_datelike:
+        # For datetime/timedelta, dtype must be specified:
+        # start and stop will first be cast to this dtype.
+        # Warning: dtype must have high enough precision so
+        # that it is at all possible to draw num equispaced
+        # samples.
+        if dtype is None:
+            raise ValueError("linspace with {} objects cannot be called with"
+                    "dtype=None".format(start.dtype))
```

Comment:
Sorry, more coming later maybe, so no need to jump on anything. A bit of nitpicking (I can fix it up also if you prefer):

The upper string needs a space at the end. Maybe `linspace()` (python does it, but I do not think we currenty do this really, so no matter). I would slightly prefer to write something like ``requires the `dtype` argument to be set to the desired output dtype.``

The ValueErrors below could also be phrased a bit nicer, but I I think they probably should just be removed or did I forget why the check was there? `linspace` with NaN returns an array full of NaN.

Reply:
Thanks for the review :)
> 
> The upper string needs a space at the end. Maybe `linspace()` (python does it, but I do not think we currenty do this really, so no matter). I would slightly prefer to write something like `` requires the `dtype` argument to be set to the desired output dtype. ``

Done.
> 
> The ValueErrors below could also be phrased a bit nicer, 

Done
> but I I think they probably should just be removed or did I forget why the check was there? `linspace` with NaN returns an array full of NaN.

If we remove those checks, we get:
```python
In [14]: np.linspace(start=np.array('2019-11-12', 'M8[s]'), 
                               stop=np.array('NaT', 'M8[s]')
                               num=3, 
                               dtype='M8[s]')                                                  
Out[14]: 
array([        '2019-11-12T00:00:00', '146138514308-05-25T07:32:16',
                               'NaT'], dtype='datetime64[s]')
```
The reason is, `stop(=NaT)` gets converted to `-9223372036854775808` via `.view(np.int64)`. 

Another fact:
```python
In [24]: np.linspace(start=np.nan, stop=100, num=3)                           
Out[24]: array([ nan,  nan, 100.])
```
What do you think should be done? 'NaT' check for `start` and `stop` or no checks?

Reply:
Hmmm, maybe we should jump through the hoops and set everything to NaT? I am not sure, an error could work for now I guess.

Reply:
I am OK with both.  If we are returning an array of `NaT`, I think for sake of consistency we should return an array of `nan` for the non-datetime equivalent case as well.

Reply:
Sorry, I need to take a step back here (and probably mainly remember all the other things that were tricky). Why can't `step` have the correct timedelta (not datetime of course) dtype? timedelta + timedelta works, timedelta * int works and datetime + timedelta works.

Sorry, this is probably just me forgetting an old discussion :/.

Reply:
@seberg : No problem, I am happy to clarify. I am not entirely sure what you mean. Is it why not do roughly 
```python
delta = stop - start # is timedelta if start/stop are datetime/timedelta
y = _nx.arange(0, num)
y *= delta
```?

If so, please take a look at [this thread](https://github.com/numpy/numpy/pull/14700#discussion_r335040226). 

The main culprit is the line `y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))`. We could do a special logic:
```python
if is_datelike:
    y = _nx.arange(0, num, dtype=_nx.integer).reshape((-1,) + (1,) * ndim(delta))
else:
    y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
```

Reply:
Yes, `y` would need special casing to ensure it is integer (in other words, it also needs to be unitless in any case). But it seems to me `delta` does not itself need to be unit-less. And if it was not, `y = y * delta` would have the unit, and the `NaT` should drop out correctly (or at least identically to `NaN`?

Reply:
You are right. `delta` does not need to be unitless and the `NaT`'s are automatically taken care of analogously to the `NaN`s. 

I got this error while trying this out:
```ipython
In [15]: np.linspace(np.array('NaT', 'M8[s]'), t2, 10,dtype='M8[D]', retstep=True)         
build/testenv/lib/python3.6/site-packages/numpy/core/function_base.py:158: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  if _nx.any(step == 0):
Out[15]: 
(array([       'NaT',        'NaT',        'NaT',        'NaT',
               'NaT',        'NaT',        'NaT',        'NaT',
               'NaT', '2019-11-12'], dtype='datetime64[D]'),
 numpy.timedelta64('NaT','s'))

In [19]: np.array('NaT', 'M8[s]') == 0                                                     
runtests.py:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  #!/usr/bin/env python
Out[19]: False
```

Any ideas what to do?

Reply:
The warning is a consequence of the `0` not having the correct dtype (datetime with unit `s`). It would go away if we do `== np.array(0, step.dtype)`. And since the ufuncs do that internally anyway, I guess the overhead, while annoying, may not matter too much.

Reply:
Ok, then. I have refactored the stuff. The testsuite passes apart from one mysterious failure in `test_all_modules_are_expected`

Reply:
@seberg : I will fix the merge conflicts and rebase to master if you are happy with the PR otherwise.

Reply:
@seberg I will rebase to master and fix the merge conflict if you are happy with the PR otherwise.